src/services/alloc/AllocService.cpp

```diff
@@ -0,0 +1,366 @@
+
+                alloc_count++;
+            }
+            hook_enabled = true;
+        }
+
+        return ret;
+    }
+
+    /**
+     * realloc
+     */
+
+    void* (*orig_realloc)(void* ptr, size_t size) = NULL;
+
+#define NUM_REALLOC_ATTRS 4
+
+    Attribute realloc_label_attr = Attribute::invalid;
+    Attribute realloc_ptr_attr = Attribute::invalid;
+    Attribute realloc_size_attr = Attribute::invalid;
+    Attribute realloc_addr_attr = Attribute::invalid;
+
+    static cali_id_t realloc_attributes[NUM_REALLOC_ATTRS] = {CALI_INV_ID};
+
+    void*
+    cali_realloc_wrapper(void *ptr, size_t size)
+    {
+        Caliper c = Caliper::sigsafe_instance();
+
+        void *ret = (*orig_realloc)(ptr, size);
+
+        if (hook_enabled) {
+            hook_enabled = false;
+            if (c) {
+                DataTracker::g_alloc_tracker.remove_allocation((uint64_t)ptr);
+                DataTracker::g_alloc_tracker.add_allocation(std::to_string(alloc_count), (uint64_t)ret, (size_t)1, {size});
+
+                Variant data[NUM_REALLOC_ATTRS];
+
+                data[0] = alloc_count;
+                data[1] = ((uint64_t)ptr);
+                data[2] = static_cast<uint64_t>(size);
+                data[3] = (uint64_t)ret;
+
+                SnapshotRecord trigger_info(NUM_REALLOC_ATTRS, realloc_attributes, data);
+                c.push_snapshot(CALI_SCOPE_PROCESS | CALI_SCOPE_THREAD, &trigger_info);
+
+                alloc_count++;
+            }
+            hook_enabled = true;
+        }
+
+        return ret;
+    }
+
+    /**
+     * free
+     */
+
+    void (*orig_free)(void* ptr) = NULL;
+
+#define NUM_FREE_ATTRS 2
+
+    Attribute free_label_attr = Attribute::invalid;
+    Attribute free_addr_attr = Attribute::invalid;
+
+    static cali_id_t free_attributes[NUM_FREE_ATTRS] = {CALI_INV_ID};
+
+    void
+    cali_free_wrapper(void *ptr)
+    {
+        Caliper c = Caliper::sigsafe_instance();
+
+        (*orig_free)(ptr);
+
+        if (hook_enabled) {
+            hook_enabled = false;
+            if (c) {
+                DataTracker::g_alloc_tracker.remove_allocation((uint64_t)ptr);
+
+                Variant data[NUM_FREE_ATTRS];
+
+                data[0] = alloc_count;
+                data[1] = ((uint64_t)ptr);
+
+                SnapshotRecord trigger_info(NUM_FREE_ATTRS, free_attributes, data);
+                c.push_snapshot(CALI_SCOPE_PROCESS | CALI_SCOPE_THREAD, &trigger_info);
+
+                alloc_count++;
+            }
+            hook_enabled = true;
+        }
+    }
+
+    void init_alloc_hooks(Caliper *c) {
+        malloc_label_attr = c->create_attribute("alloc.malloc.label", CALI_TYPE_UINT, CALI_ATTR_DEFAULT);
+        malloc_size_attr = c->create_attribute("alloc.malloc.size", CALI_TYPE_UINT, CALI_ATTR_DEFAULT);
+        malloc_addr_attr = c->create_attribute("alloc.malloc.address", CALI_TYPE_UINT, CALI_ATTR_DEFAULT);
```

Comment:
Wouldn't it make more sense to have a `alloc.fn` attribute which is set to `malloc`, `calloc`, or `realloc` rather than individual size/label/address attributes for each allocation function? I think this would make some analyses easier as they can then be function agnostic, and it would also be easier to add additional alloc functions (like `new`)

Reply:
I did this because they have different arguments, i.e. calloc has an element size and a number of elements, rather than just a number of bytes, but I like having `alloc.fn`, I'll figure out a compromise